File: apps/presence/test/presence_web/plugs/origin_guard_test.exs
defmodule PresenceWeb.Plugs.OriginGuardTest do
  use PresenceWeb.ConnCase, async: false

  alias PresenceWeb.Plugs.OriginGuard

  describe "allowed_origin?/1" do
    test "allows default web and native origins" do
      assert OriginGuard.allowed_origin?("http://localhost:3000")
      assert OriginGuard.allowed_origin?("http://127.0.0.1:3000")
      assert OriginGuard.allowed_origin?("tauri://localhost")
      assert OriginGuard.allowed_origin?("http://tauri.localhost")
      assert OriginGuard.allowed_origin?("https://tauri.localhost")
    end

    test "allows loopback aliases for local development" do
      assert OriginGuard.allowed_origin?("http://0.0.0.0:3000")
    end

    test "blocks untrusted origins" do
      refute OriginGuard.allowed_origin?("http://evil.com")
      refute OriginGuard.allowed_origin?("https://attacker.org")
      refute OriginGuard.allowed_origin?(nil)
    end
  end

  describe "call/2 plug verification" do
    setup do
      Application.put_env(:presence, :env, :dev)
      Application.put_env(:presence, :internal_api_key, "test-internal-key")

      on_exit(fn ->
        Application.put_env(:presence, :env, :test)
        Application.delete_env(:presence, :internal_api_key)
      end)

      :ok
    end

    test "allows requests with valid origin header" do
      conn =
        build_conn()
        |> put_req_header("origin", "http://localhost:3000")
        |> OriginGuard.call([])

      refute conn.halted
      assert conn.status != 403
    end

    test "allows requests with valid referer header" do
      conn =
        build_conn()
        |> put_req_header("referer", "http://localhost:3000/boards/ws-1")
        |> OriginGuard.call([])

      refute conn.halted
      assert conn.status != 403
    end

    test "blocks requests with untrusted origin" do
      conn =
        build_conn()
        |> put_req_header("origin", "http://evil.com")
        |> OriginGuard.call([])

      assert conn.halted
      assert conn.status == 403
    end

    test "blocks direct browser navigation requests (sec-fetch-dest: document)" do
      conn =
        build_conn()
        |> put_req_header("sec-fetch-dest", "document")
        |> put_req_header("sec-fetch-mode", "navigate")
        |> OriginGuard.call([])

      assert conn.halted
      assert conn.status == 403
    end

    test "allows requests with internal API key in authorization header" do
      conn =
        build_conn()
        |> put_req_header("authorization", "Bearer test-internal-key")
        |> OriginGuard.call([])

      refute conn.halted
      assert conn.status != 403
    end

    test "allows requests with internal API key in x-internal-key header" do
      conn =
        build_conn()
        |> put_req_header("x-internal-key", "test-internal-key")
        |> OriginGuard.call([])

      refute conn.halted
      assert conn.status != 403
    end

    test "allows OPTIONS CORS preflight requests" do
      conn =
        build_conn(:options, "/health")
        |> OriginGuard.call([])

      refute conn.halted
      assert conn.status != 403
    end
  end
end

File: apps/workers/src/middleware/origin-guard.test.ts
import { describe, expect, it } from "bun:test";

process.env.DATABASE_URL = "postgres://dummy";
process.env.NODE_ENV = "development";
process.env.WEB_URL = "http://localhost:3000";
process.env.BETTER_AUTH_URL = "http://localhost:3002";
process.env.BETTER_AUTH_SECRET = "test-secret-must-be-21-chars-long!!";
process.env.GITHUB_CLIENT_ID = "test-github-client-id";
process.env.GITHUB_CLIENT_SECRET = "test-github-client-secret";
process.env.JWKS_ENCRYPTION_KEY = "test-jwks-encryption-key-32chars!!";
process.env.REDIS_URL = "redis://localhost:6379";
process.env.INTERNAL_API_KEY = "test-internal-key";

const { Elysia } = await import("elysia");
const { isAllowedOrigin, originGuard } = await import("./origin-guard");

describe("isAllowedOrigin", () => {
  it("allows exact match for default origins", () => {
    expect(isAllowedOrigin("http://localhost:3000")).toBe(true);
    expect(isAllowedOrigin("http://127.0.0.1:3000")).toBe(true);
    expect(isAllowedOrigin("tauri://localhost")).toBe(true);
    expect(isAllowedOrigin("http://tauri.localhost")).toBe(true);
    expect(isAllowedOrigin("https://tauri.localhost")).toBe(true);
  });

  it("allows configured custom origins", () => {
    expect(
      isAllowedOrigin("https://app.lumen.dev", ["https://app.lumen.dev"])
    ).toBe(true);
  });

  it("allows loopback aliases for local dev", () => {
    expect(isAllowedOrigin("http://0.0.0.0:3000")).toBe(true);
  });

  it("blocks untrusted external origins", () => {
    expect(isAllowedOrigin("http://evil.com")).toBe(false);
    expect(isAllowedOrigin("https://attacker.org")).toBe(false);
    expect(isAllowedOrigin(null)).toBe(false);
    expect(isAllowedOrigin(undefined)).toBe(false);
  });
});

describe("originGuard middleware", () => {
  const createTestApp = () =>
    new Elysia()
      .use(originGuard)
      .get("/test", () => ({ success: true }))
      .get("/api/auth/jwks", () => ({ keys: [] }))
      .get("/api/auth/callback/github", () => ({ callback: true }));

  it("allows requests from allowed origin", async () => {
    const app = createTestApp();
    const res = await app.handle(
      new Request("http://localhost/test", {
        headers: {
          origin: "http://localhost:3000",
        },
      })
    );

    expect(res.status).toBe(200);
    const data = (await res.json()) as { success: boolean };
    expect(data.success).toBe(true);
  });

  it("allows requests with allowed referer header", async () => {
    const app = createTestApp();
    const res = await app.handle(
      new Request("http://localhost/test", {
        headers: {
          referer: "http://localhost:3000/boards/ws-1",
        },
      })
    );

    expect(res.status).toBe(200);
  });

  it("blocks requests with untrusted origin", async () => {
    const app = createTestApp();
    const res = await app.handle(
      new Request("http://localhost/test", {
        headers: {
          origin: "http://evil.com",
        },
      })
    );

    expect(res.status).toBe(403);
    const data = (await res.json()) as { error: string };
    expect(data.error).toBe("Forbidden");
  });

  it("blocks direct browser navigation requests (sec-fetch-dest: document)", async () => {
    const app = createTestApp();
    const res = await app.handle(
      new Request("http://localhost/test", {
        headers: {
          "sec-fetch-dest": "document",
          "sec-fetch-mode": "navigate",
        },
      })
    );

    expect(res.status).toBe(403);
    const data = (await res.json()) as { error: string; message: string };
    expect(data.message).toBe("Direct browser access is blocked");
  });

  it("allows OAuth callback endpoints even on browser navigation", async () => {
    const app = createTestApp();
    const res = await app.handle(
      new Request("http://localhost/api/auth/callback/github", {
        headers: {
          "sec-fetch-dest": "document",
          "sec-fetch-mode": "navigate",
        },
      })
    );

    expect(res.status).toBe(200);
  });

  it("allows /api/auth/jwks endpoint without Origin header for JWT verification", async () => {
    const app = createTestApp();
    const res = await app.handle(
      new Request("http://localhost/api/auth/jwks", {
        method: "GET",
      })
    );

    expect(res.status).toBe(200);
    const data = (await res.json()) as { keys: unknown[] };
    expect(data.keys).toBeDefined();
  });

  it("allows internal service calls with x-internal-key", async () => {
    const app = createTestApp();
    const res = await app.handle(
      new Request("http://localhost/test", {
        headers: {
          "x-internal-key": "test-internal-key",
        },
      })
    );

    expect(res.status).toBe(200);
  });

  it("allows CORS OPTIONS preflight", async () => {
    const app = createTestApp();
    const res = await app.handle(
      new Request("http://localhost/test", {
        method: "OPTIONS",
      })
    );

    expect(res.status).not.toBe(403);
  });
});
